### PR TITLE
Update range max of laser scan with max from point cloud

### LINKED ISSUE #148: laser scan seems off at higher distances.
image

As you can see in the figure (topdown ortho view in RVIZ) the laserscan does not match the pointcloud at further distances.

### PR DESCRIPTION
**Pull request by PerrineAguiar**
This PR updates range max of laser scan with max from point cloud.
It also performs some small refactoring.

### CODE REVIEW

File: tango_ros_common/tango_ros_native/src/tango_ros_node.cpp [hunk 1]
```diff
@@ -158,19 +161,24 @@ void toLaserScanRange(double x, double y, double z, double min_height,
...
+                 geometry_msgs::Transform point_cloud_T_laser_message,
+                 sensor_msgs::LaserScan* laser_scan) {
   for (size_t i = 0; i < tango_point_cloud.num_points; ++i) {
-    // Laser scan frame is rotated of 90 degrees around x axis with respect to
-    // point cloud frame.
-    double x = tango_point_cloud.points[i][0];
-    double y = tango_point_cloud.points[i][2];
-    double z = -tango_point_cloud.points[i][1];
-    toLaserScanRange(x, y, z, min_height, max_height, laser_scan);
+    tf::Vector3 point_cloud_p(tango_point_cloud.points[i][0],
```
Comment (mcopejans) [inline tango_ros_common/tango_ros_native/src/tango_ros_node.cpp]: const

File: tango_ros_common/tango_ros_native/src/tango_ros_node.cpp [hunk 2]
```diff
@@ -158,19 +161,24 @@ void toLaserScanRange(double x, double y, double z, double min_height,
...
 // @param laser_scan, the output LaserScan.
 void toLaserScan(const TangoPointCloud& tango_point_cloud,
-                   double time_offset,
-                   double min_height,
-                   double max_height,
-                   sensor_msgs::LaserScan* laser_scan) {
+                 double time_offset,
+                 double min_height,
+                 double max_height,
+                 geometry_msgs::Transform point_cloud_T_laser_message,
```
Comment (mcopejans) [inline tango_ros_common/tango_ros_native/src/tango_ros_node.cpp]: const ref

File: tango_ros_common/tango_ros_native/src/tango_ros_node.cpp [hunk 3]
```diff
@@ -158,19 +161,24 @@ void toLaserScanRange(double x, double y, double z, double min_height,
 // included in laser scan.
 // @param max_height maximum height for a point of the point cloud to be
 // included in laser scan.
+// @param point_cloud_T_laser_message transform message between point cloud
+// frame and laser scan frame.
```
Comment (mcopejans) [inline tango_ros_common/tango_ros_native/src/tango_ros_node.cpp]: More correct: Transformation from point cloud frame to laser scan frame. Or transform of laser scan wrt. point cloud.

File: tango_ros_common/tango_ros_native/src/tango_ros_node.cpp [hunk 4]
```diff
@@ -158,19 +161,24 @@ void toLaserScanRange(double x, double y, double z, double min_height,
...
-    double x = tango_point_cloud.points[i][0];
-    double y = tango_point_cloud.points[i][2];
-    double z = -tango_point_cloud.points[i][1];
-    toLaserScanRange(x, y, z, min_height, max_height, laser_scan);
+    tf::Vector3 point_cloud_p(tango_point_cloud.points[i][0],
+                              tango_point_cloud.points[i][1],
+                              tango_point_cloud.points[i][2]);
+    tf::Transform point_cloud_T_laser;
+    tf::transformMsgToTF(point_cloud_T_laser_message, point_cloud_T_laser);
+    tf::Vector3 laser_p = point_cloud_T_laser.inverse() * point_cloud_p;
```
Comment (mcopejans) [inline tango_ros_common/tango_ros_native/src/tango_ros_node.cpp]: Maybe cleaner to make this a helper function transformPointCloud

**Comment by PerrineAguiar** [comment]
@mcopejans Thanks for the quick review.
The laser scan frame is rotated wrt. the point cloud frame to follow the ROS convention which measures laser scan angles around the Z-axis (see http://docs.ros.org/api/sensor_msgs/html/msg/LaserScan.html).
I added a comment in the code to explain this.

File: tango_ros_common/tango_ros_native/src/tango_ros_node.cpp [hunk 5]
```diff
@@ -112,6 +112,17 @@ void toPointCloud2(const TangoPointCloud& tango_point_cloud,
   }
   point_cloud->header.stamp.fromSec((tango_point_cloud.timestamp + time_offset) / 1e3);
 }
+// Transform a point from frame 1 to frame2.
+// @param frame1_p point expressed in frame1
+// @param frame1_T_frame2_message transform message from frame1 to frame2.
+// @param frame2_p output point expressed in frame2.
+void transformPoint(const tf::Vector3& frame1_p,
+                    const geometry_msgs::Transform& frame1_T_frame2_message,
```
Comment (mcopejans) [inline tango_ros_common/tango_ros_native/src/tango_ros_node.cpp]: Would be better if the function directly takes tf::Transform as input. This avoids converting the message type to a transform type for every point in the point cloud.

**Comment by PerrineAguiar** [comment]
@mcopejans Thanks, I addressed your last comment.

**Comment by PerrineAguiar** [comment]
@smits @mcopejans I think we can merge this PR.
We can test if issue #148 appears again and fix it if necessary in another PR.

**Pull request merged**